### ISSUE TITLE
Not actually an issue just a proposal

### ISSUE THREAD

**Issue by Hizoul** [staydecent/nativescript-preact#4]
since I didn't find a personal message button I'll just share this here:
https://github.com/Hizoul/proposal-for-preact-to-nativescript

**Comment by staydecent** [comment]
Definitely interesting. And, I think may open up possibility for NS specific optimizations.

I wasn't sure how to approach this, but eventually ended up following the advice of @_developit and just using undom as that allows Preact to work out of the box in any JS environment.

I'm curious the advantages you're seeking?

**Comment by Hizoul** [comment]
The main reasoning behind this is to eliminate uneeded code and possible bottlenecks.

Your Approach: Preact translates JSX to Undoms Emulated web-dom which is then translated to nativescripts "dom" based on the emulated web-dom.

This Approach: Directly translate JSX  into NativeScripts DOM. The lib could then theoretically like the [ink](https://github.com/vadimdemedes/ink) project be completely independt from preact, and not rely on it at all if wanted.

Advantages: No middle layer, and direct translation of JSX-Elements.

I do understand wanting to just reuse preact completely and not think about the rendering / VDOM traversal etc. I believe work could be done to improve isolating preacts JSX-Code from webdom-creation code so that it can more easily be applied to render different stuff. Especially looking at preact-to-server etc., they replicate many needed parts of VDOM traversal code to achieve what they want from the original preact code.

**Comment by staydecent** [comment]
Ah I see. My goal is just to enable the (P)React pattern of building UIs with components all the way down. 

This approach seems like the least amount of added maintenance overhead. Any improvements to Preact can instantly be used in the NS environment. And because there are multiple well-made implementations of DOM api in JS, there's not much risk there either. I chose Undom for it's small size.

I think the translation layer of DOM events/apis to NS events is pretty simple, and will actually allow this library to remain quite tiny.

The less code introduced between Preact and NS, the less new surface area to introduce bugs.

Anyway, JSX to NS sounds very cool and I'll be interested to see where you take it :)

**Comment by developit** [comment]
Chiming in here - stripping out Undom seems fine, the effect would essentially be that you're setting up DOM proxy methods rather than objects. That would allow preact to treat NativeScript's view/widget instance as if it were a DOM.

Here's some untested stuff that popped into my head:

```js
import { Text } from 'tns-core-modules/ui/text-view';
let types = {};

// stuff to mix into NS's view prototypes
let extensions = {
  setAttribute(name, value) {
    this.set(name, value);
  },
  removeAttribute(name) {
    this.set(name, null);
  },
  appendChild(child) {
    if ('text' in this && child.splitText!=null) {
      this.text = child.nodeValue;
    }
    else {
      this.addChild(child);
    }
  },
  insertBefore(child, ref) {
    // find the index at which to insert the child based on ref:
    let offset, index = -1;
    this.eachChild( c => {
      index++;
      if (c===ref) offset = index;
    });
    if (offset!=null) {
      this._addView(child, offset);
    }
    else {
      this.addChild(child);
    }
  },
  removeChild(child) {
    if ('text' in this && child.splitText!=null) {
      this.text = '';
    }
    else {
      this._removeView(child);
    }
  }
};

let document = {
  createElement(type) {
    // imports and augments NS view classes on first use
    type = type.toLowerCase();
    if (type in types) return types[type];
    let m = require('tns-core-modules/ui/'+type);
    let el;
    // find matching named export:
    for (let i in m) if (i.toLowerCase()===type) {
      el = m[i];
      break;
    }
    Object.assign(el.prototype, extensions);
    return types[type] = el;
  },
  createTextNode(text) {
    let el = new Text();
    el.text = text;
    Object.defineProperty(el, 'nodeValue', {
      set(v) { this.text = v },
      get() { return this.text }
    });
    el.splitText = () => null;
    return el;
  }
};

That said, this does away with the automatic batching undom's simple MutationObserver approach does for us.  Performance-wise, undom has very little overhead - it's not a real DOM, just a basic object graph.  It's likely the MutationObserver goodness outweighs any cost of having a DOM structure around, plus the added benefit of having a standardized layer in front of NativeScript to simplify syncing properties over the boundary.

**Comment by Hizoul** [comment]
Sorry for the extremely late answer. I finally found a bunch of free time to give your suggestion a try.
It worked mostly out of the box with minor adjustments.
Within preact itself I merely had to slightly adjust event handling a little to make it work with NativeScript (see https://github.com/Hizoul/proposal-for-preact-to-nativescript/commit/e5fd0e3f62debad5e55dd99f9ac998399e83220f)

It seems though as I am somehow doing something wrong within createElement or so, because it always does a full rerender on change of text. Also don't know if related but switching to the second page and triggering changes there removes the StackLayout element but never adds anything back haven't figured out why yet either :/
// Edit:
i'll study the preact [internals](https://medium.com/@asolove/preact-internals-1-the-easy-parts-3a081fa36205) [blog](https://medium.com/@asolove/preact-internals-2-the-component-model-36a05e32957b) [series](https://medium.com/@asolove/preact-internals-3-some-fiddly-little-bits-f353b1ad7abc) and see if it'll help me figure this out
// Edit2: I found the problem that was responsible for both issues. I added twice to the childNodes array hence preact correctly thought it needed to recreate the component :) will update upon further interesting findings :)

**Comment by Hizoul** [comment]
I will close this issue because the the [proposal-for-preact-to-nativescript](https://github.com/Hizoul/proposal-for-preact-to-nativescript)  developed into something that can render all basic nativescript components in the demoApp of the repo. I extracted the renderer into an experimental library [preact-to-nativescript](https://github.com/Hizoul/preact-to-nativescript).
I also submitted a [pull request to preact](https://github.com/developit/preact/pull/978) for the mentioned event handling adjustments.
